### ISSUE TITLE
Pull dependencies from Clojars?

### ISSUE THREAD

**Issue by cjsauer** [arcadia-unity/Arcadia#245]
Is Clojars supported via the `:dependencies` option? It seems like I can pull Maven dependencies ok, but my Clojars dependency is throwing this:

```
System.ArgumentNullException: Argument cannot be null.
Parameter name: address
at System.Net.WebClient.DownloadString (string) <IL 0x00010, 0x000a9>
at (wrapper remoting-invoke-with-check) System.Net.WebClient.DownloadString (string) <IL 0x00033, 0x00141>
at (wrapper dynamic-method) object.CallSite.Target (System.Runtime.CompilerServices.Closure,System.Runtime.CompilerServices.CallSite,object,object) <IL 0x00048, 0x00290>
at arcadia/packages$download_string__40935.__interop_DownloadString40937 (object,object) <IL 0x0000f, 0x0008c>
at arcadia/packages$download_string__40935.invokeStatic (object) <IL 0x00008, 0x000de>
at arcadia/packages$download_string__40935.invoke (object) <IL 0x00006, 0x0007e>
at arcadia/packages$dependencies__41051.invokeStatic (object) <IL 0x00112, 0x00ea4>
at arcadia/packages$dependencies__41051.invoke (object) <IL 0x00006, 0x0007e>
at arcadia/packages$all_dependencies__41067.invokeStatic (object) <IL 0x000d3, 0x00a53>
at arcadia/packages$all_dependencies__41067.invoke (object) <IL 0x00006, 0x0007e>
at arcadia/packages$install__41275.invokeStatic (object,object) <IL 0x00226, 0x01f05>
at arcadia/packages$install__41275.invoke (object,object) <IL 0x0000c, 0x000b4>
at arcadia/packages$install_step__41311.invokeStatic (object) <IL 0x0003c, 0x0041a>
at arcadia/packages$install_step__41311.invoke (object) <IL 0x00006, 0x0007e>
at clojure/core$mapfn__2577fn__2581__2588.invoke (object,object) <IL 0x00018, 0x00231>
at clojure/core/protocols$eval__10177fn__10182__10186.invoke (object,object,object) <IL 0x000a7, 0x00a16>
at clojure/core/protocols$eval__9927fn__9932G__9896__9949__9952.invoke (object,object,object) <IL 0x00057, 0x003e7>
at clojure/core/protocols$seq_reduce__9979.invokeStatic (object,object,object) <IL 0x00055, 0x00481>
at clojure/core/protocols$seq_reduce__9979.invoke (object,object,object) <IL 0x00012, 0x000ea>
at clojure/core/protocols$eval__10060fn__10065__10070.invoke (object,object,object) <IL 0x00012, 0x0016a>
at clojure/core/protocols$eval__9844fn__9849G__9813__9871__9874.invoke (object,object,object) <IL 0x00057, 0x003e7>
at clojure/core$transduce__11081.invokeStatic (object,object,object,object) <IL 0x00076, 0x008b3>
at clojure/core$transduce__11081.invoke (object,object,object,object) <IL 0x00019, 0x0011c>
at clojure/core$into__11088.invokeStatic (object,object,object) <IL 0x0005c, 0x00711>
at clojure/core$into__11088.invoke (object,object,object) <IL 0x00012, 0x000ea>
at arcadia/packages$install_1__41378.invokeStatic () <IL 0x001ef, 0x01bd4>
at arcadia/packages$install_1__41378.invoke () <IL 0x00000, 0x00048>
at arcadia/packages$install_all_depsfn__41391__41395.invoke () <IL 0x0000f, 0x00176>

UnityEngine.Debug:Log(Object)
arcadia/packages$install_all_depsfn__41391__41395:invoke()
arcadia/internal/thread$start_threadfn__23638__23642:invoke()
System.Object:lambda_method(Closure)
```

**Comment by cjsauer** [comment]
Took a look at the source code in the meantime. In the `packages.clj` source file, I see [these two functions](https://github.com/arcadia-unity/Arcadia/blob/develop/Source/arcadia/packages.clj#L163-L167) are only picking off the `(first)` from urls: 

```Clojure
(defn jar-url [group artifact version]
  (first (jar-urls group artifact version)))

(defn pom-url [group artifact version]
  (first (pom-urls group artifact version)))
```

where urls is ultimately sourced from [this definition](https://github.com/arcadia-unity/Arcadia/blob/develop/Source/arcadia/packages.clj#L81-L86):

```Clojure
(def url-prefixes
  ["http://central.maven.org/maven2/"
   "http://search.maven.org/remotecontent?filepath="
   "https://oss.sonatype.org/content/repositories/snapshots/"
   "https://repo1.maven.org/maven2/"
   "https://clojars.org/repo/"])
```

Does this mean that multiple repositories hasn't been implemented yet, or is there some configuration setup that allows me to choose which repo I'd like to source from?

**Comment by cjsauer** [comment]
I've tried moving the clojars url to be the first in the above `urls` definition, and I'm still receiving the same error unfortunately.

**Comment by cjsauer** [comment]
I believe I have the issue pinpointed to the `url-exists?` function. It appears that the package I want to depend on doesn't have a `maven-metadata.xml` file available. Is anyone familiar with how to publish this file to Clojars?

**Comment by cjsauer** [comment]
Making more headway on this. It appears that a lot of the repository urls use the `https` protocol, which is resulting in the error, status code 405: 

```Unhandled Exception: System.Net.WebException: Error getting response stream (Write: The authentication or decryption has failed.)```

Within the `url-exists?` [function](https://github.com/arcadia-unity/Arcadia/blob/develop/Source/arcadia/packages.clj#L88-L96) Arcadia is returning false on _any_ exception. This isn't correct, because it is catching these 405 errors and assuming the file doesn't exist, when actually it's the SSL certificate that is causing the exception.

By consulting [this StackOverflow post](https://stackoverflow.com/questions/4926676/mono-https-webrequest-fails-with-the-authentication-or-decryption-has-failed) I added in the `MyRemoteCertificateValidationCallback` method to the `Initialize.cs` script within Arcadia. Now, all of the aforementioned 405 status code messages have disappeared, yet I'm still unable to successfully install packages. I'm now receiving this error in the Unity console:

```
System.ArgumentNullException: Argument cannot be null.
Parameter name: s
  at System.IO.StringReader..ctor (System.String s) [0x0000c] in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.IO/StringReader.cs:51 
  at (wrapper remoting-invoke-with-check) System.IO.StringReader:.ctor (string)
  at arcadia/packages$read_xml__32848.invokeStatic (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$read_xml__32848.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$dependencies__33034.invokeStatic (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$dependencies__33034.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$mapfn__2595__2599.invoke () [0x00000] in <filename unknown>:0 
  at clojure.lang.LazySeq.sval () [0x00000] in <filename unknown>:0 
  at (wrapper synchronized) clojure.lang.LazySeq:sval ()
  at clojure.lang.LazySeq.seq () [0x00000] in <filename unknown>:0 
  at (wrapper synchronized) clojure.lang.LazySeq:seq ()
  at clojure.lang.RT.seq (System.Object coll) [0x00000] in <filename unknown>:0 
  at clojure/core$seq__80__83.invokeStatic (System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$seq__80__83.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$apply__501.invokeStatic (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$apply__501.invoke (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$mapcat__2674.invokeStatic (System.Object , ISeq ) [0x00000] in <filename unknown>:0 
  at clojure/core$mapcat__2674.doInvoke (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure.lang.RestFn.invoke (System.Object arg1, System.Object arg2) [0x00000] in <filename unknown>:0 
  at arcadia/packages$all_dependenciesfn__28452__28456.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at clojure.lang.Iterate.first () [0x00000] in <filename unknown>:0 
  at clojure.lang.RT.first (System.Object x) [0x00000] in <filename unknown>:0 
  at clojure/core$first__26__29.invokeStatic (System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$first__26__29.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$take_whilefn__2794__2798.invoke () [0x00000] in <filename unknown>:0 
  at clojure.lang.LazySeq.sval () [0x00000] in <filename unknown>:0 
  at (wrapper synchronized) clojure.lang.LazySeq:sval ()
  at clojure.lang.LazySeq.seq () [0x00000] in <filename unknown>:0 
  at (wrapper synchronized) clojure.lang.LazySeq:seq ()
  at clojure.lang.Cons.next () [0x00000] in <filename unknown>:0 
  at clojure.lang.RT.BoundedLength (ISeq list, Int32 limit) [0x00000] in <filename unknown>:0 
  at clojure.lang.RestFn.applyTo (ISeq arglist) [0x00000] in <filename unknown>:0 
  at clojure/core$apply__501.invokeStatic (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$apply__501.invoke (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$all_dependencies__28459.invokeStatic (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$all_dependencies__28459.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$install__28667.invokeStatic (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$install__28667.invoke (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$install_step__28703.invokeStatic (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$install_step__28703.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$mapfn__2577fn__2581__2588.invoke (System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core/protocols$eval__10177fn__10182__10186.invoke (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core/protocols$eval__9927fn__9932G__9896__9949__9952.invoke (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core/protocols$seq_reduce__9979.invokeStatic (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core/protocols$seq_reduce__9979.invoke (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core/protocols$eval__10060fn__10065__10070.invoke (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core/protocols$eval__9844fn__9849G__9813__9871__9874.invoke (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$transduce__11081.invokeStatic (System.Object , System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$transduce__11081.invoke (System.Object , System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$into__11088.invokeStatic (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at clojure/core$into__11088.invoke (System.Object , System.Object , System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/packages$install_1__28770.invokeStatic () [0x00000] in <filename unknown>:0 
  at arcadia/packages$install_1__28770.invoke () [0x00000] in <filename unknown>:0 
  at arcadia/packages$install_all_depsfn__28783__28787.invoke () [0x00000] in <filename unknown>:0 
UnityEngine.Debug:Log(Object)
arcadia/packages$install_all_depsfn__28783__28787:invoke()
arcadia/internal/thread$start_threadfn__23638__23642:invoke()
System.Object:lambda_method(Closure)
```

**Comment by selfsame** [comment]
@cjsauer yep windows and linux can't handle the `https` links, there's an issue at https://github.com/arcadia-unity/Arcadia/issues/175, but in the meantime i think if you change  the clojars to `http` and remove the other `https` repositories it works

**Comment by cjsauer** [comment]
@selfsame thanks for the reply. Unfortunately I'm still getting the following error: 

```
System.ArgumentNullException: Argument cannot be null.
Parameter name: address
at System.Net.WebClient.DownloadString (string) <IL 0x00010, 0x000a9>
at (wrapper remoting-invoke-with-check) System.Net.WebClient.DownloadString (string) <IL 0x00033, 0x00141>
at (wrapper dynamic-method) object.CallSite.Target (System.Runtime.CompilerServices.Closure,System.Runtime.CompilerServices.CallSite,object,object) <IL 0x00048, 0x00290>
at (wrapper dynamic-method) object.CallSite.Target (System.Runtime.CompilerServices.Closure,System.Runtime.CompilerServices.CallSite,object,object) <IL 0x00137, 0x009dd>
at (wrapper dynamic-method) object.CallSite.Target (System.Runtime.CompilerServices.Closure,System.Runtime.CompilerServices.CallSite,object,object) <IL 0x00086, 0x00420>
at arcadia/packages$download_string__28327.__interop_DownloadString28329 (object,object) <IL 0x0000f, 0x0008c>
at arcadia/packages$download_string__28327.invokeStatic (object) <IL 0x00008, 0x000f3>
at arcadia/packages$download_string__28327.invoke (object) <IL 0x00006, 0x0007e>
at arcadia/packages$dependencies__28443.invokeStatic (object) <IL 0x00112, 0x00ea4>
at arcadia/packages$dependencies__28443.invoke (object) <IL 0x00006, 0x0007e>
at clojure/core$mapfn__2595__2599.invoke () <IL 0x00108, 0x00e4c>
at clojure.lang.LazySeq.sval () <IL 0x00012, 0x0007f>
at (wrapper synchronized) clojure.lang.LazySeq.sval () <IL 0x00009, 0x000a7>
at clojure.lang.LazySeq.seq () <IL 0x00001, 0x00054>
at (wrapper synchronized) clojure.lang.LazySeq.seq () <IL 0x00009, 0x000a7>
at clojure.lang.RT.seq (object) <IL 0x0001d, 0x00181>
at clojure/core$seq__80__83.invokeStatic (object) <IL 0x00006, 0x0007a>
at clojure/core$seq__80__83.invoke (object) <IL 0x00006, 0x0007e>
at clojure/core$apply__501.invokeStatic (object,object) <IL 0x00016, 0x00255>
at clojure/core$apply__501.invoke (object,object) <IL 0x0000c, 0x000b4>
at clojure/core$mapcat__2674.invokeStatic (object,clojure.lang.ISeq) <IL 0x00039, 0x00300>
at clojure/core$mapcat__2674.doInvoke (object,object) <IL 0x0000c, 0x00164>
at clojure.lang.RestFn.invoke (object,object) <IL 0x0006d, 0x003bf>
at arcadia/packages$all_dependenciesfn__28452__28456.invoke (object) <IL 0x0001a, 0x0018f>
at clojure.lang.Iterate.first () <IL 0x0001f, 0x00094>
at clojure.lang.RT.first (object) <IL 0x0001d, 0x0020d>
at clojure/core$first__26__29.invokeStatic (object) <IL 0x00006, 0x0007a>
at clojure/core$first__26__29.invoke (object) <IL 0x00006, 0x0007e>
at clojure/core$take_whilefn__2794__2798.invoke () <IL 0x0005c, 0x005a6>
at clojure.lang.LazySeq.sval () <IL 0x00012, 0x0007f>
at (wrapper synchronized) clojure.lang.LazySeq.sval () <IL 0x00009, 0x000a7>
at clojure.lang.LazySeq.seq () <IL 0x00001, 0x00054>
at (wrapper synchronized) clojure.lang.LazySeq.seq () <IL 0x00009, 0x000a7>
at clojure.lang.Cons.next () <IL 0x00006, 0x00075>
at clojure.lang.RT.next (object) <IL 0x0001d, 0x0020d>
at clojure/core$next__32__35.invokeStatic (object) <IL 0x00006, 0x0007a>
at clojure/core$next__32__35.invoke (object) <IL 0x00006, 0x0007e>
at clojure/core$concatcat__605fn__608__612.invoke () <IL 0x0018c, 0x0159d>
at clojure.lang.LazySeq.sval () <IL 0x00012, 0x0007f>
at (wrapper synchronized) clojure.lang.LazySeq.sval () <IL 0x00009, 0x000a7>
at clojure.lang.LazySeq.seq () <IL 0x00001, 0x00054>
at (wrapper synchronized) clojure.lang.LazySeq.seq () <IL 0x00009, 0x000a7>
at clojure.lang.ChunkedCons.chunkedNext () <IL 0x00006, 0x00073>
at clojure/core$chunk_next__563.invokeStatic (object) <IL 0x00006, 0x00131>
at clojure/core$chunk_next__563.invoke (object) <IL 0x00006, 0x0007e>
at clojure/core/protocols$eval__10149fn__10154__10158.invoke (object,object,object) <IL 0x000dd, 0x00cd7>
at clojure/core/protocols$eval__9927fn__9932G__9896__9949__9952.invoke (object,object,object) <IL 0x00057, 0x003e7>
at clojure/core/protocols$seq_reduce__9979.invokeStatic (object,object,object) <IL 0x00055, 0x00481>
at clojure/core/protocols$seq_reduce__9979.invoke (object,object,object) <IL 0x00012, 0x000ea>
at clojure/core/protocols$eval__10060fn__10065__10070.invoke (object,object,object) <IL 0x00012, 0x0016a>
at clojure/core/protocols$eval__9844fn__9849G__9813__9871__9874.invoke (object,object,object) <IL 0x00057, 0x003e7>
at clojure/core$reduce__11001.invokeStatic (object,object,object) <IL 0x00069, 0x00716>
at clojure/core$reduce__11001.invoke (object,object,object) <IL 0x00012, 0x000ea>
at clojure/core$into__11088.invokeStatic (object,object) <IL 0x0005b, 0x006ed>
at clojure/core$into__11088.invoke (object,object) <IL 0x0000c, 0x000b4>
at arcadia/packages$all_dependencies__28459.invokeStatic (object) <IL 0x000e7, 0x00b1f>
at arcadia/packages$all_dependencies__28459.invoke (object) <IL 0x00006, 0x0007e>
at arcadia/packages$install__28667.invokeStatic (object,object) <IL 0x00226, 0x01f05>
at arcadia/packages$install__28667.invoke (object,object) <IL 0x0000c, 0x000b4>
at arcadia/packages$install_step__28703.invokeStatic (object) <IL 0x0003c, 0x0041a>
at arcadia/packages$install_step__28703.invoke (object) <IL 0x00006, 0x0007e>
at clojure/core$mapfn__2577fn__2581__2588.invoke (object,object) <IL 0x00018, 0x00231>
at clojure/core/protocols$eval__10177fn__10182__10186.invoke (object,object,object) <IL 0x000a7, 0x00a16>
at clojure/core/protocols$eval__9927fn__9932G__9896__9949__9952.invoke (object,object,object) <IL 0x00057, 0x003e7>
at clojure/core/protocols$seq_reduce__9979.invokeStatic (object,object,object) <IL 0x00055, 0x00481>
at clojure/core/protocols$seq_reduce__9979.invoke (object,object,object) <IL 0x00012, 0x000ea>
at clojure/core/protocols$eval__10060fn__10065__10070.invoke (object,object,object) <IL 0x00012, 0x0016a>
at clojure/core/protocols$eval__9844fn__9849G__9813__9871__9874.invoke (object,object,object) <IL 0x00057, 0x003e7>
at clojure/core$transduce__11081.invokeStatic (object,object,object,object) <IL 0x00076, 0x008b3>
at clojure/core$transduce__11081.invoke (object,object,object,object) <IL 0x00019, 0x0011c>
at clojure/core$into__11088.invokeStatic (object,object,object) <IL 0x0005c, 0x00711>
at clojure/core$into__11088.invoke (object,object,object) <IL 0x00012, 0x000ea>
at arcadia/packages$install_1__28770.invokeStatic () <IL 0x001ef, 0x01bd4>
at arcadia/packages$install_1__28770.invoke () <IL 0x00000, 0x00048>
at arcadia/packages$install_all_depsfn__28783__28787.invoke () <IL 0x0000f, 0x00176>

UnityEngine.Debug:Log(Object)
arcadia/packages$install_all_depsfn__28783__28787:invoke()
arcadia/internal/thread$start_threadfn__23638__23642:invoke()
System.Object:lambda_method(Closure)
```

**Comment by selfsame** [comment]
could be either 
* one of those links is invalid (some of the `https` ones don't allow http, and i think that breaks things)
* issues with the transitive deps, I wasn't able to get one working because it depended on clojurescript, you could try `[selfsame/soa "0.6-SNAPSHOT"]` which doesn't have any secondary deps

**Comment by nasser** [comment]
@cjsauer i have a fix in the works. what OS are you on? can you try and eval this gist? https://gist.github.com/nasser/306f016cfec3c5f9c01e1fd85c7da67a

**Comment by cjsauer** [comment]
@nasser I'm on Windows. Where should I evaluate that gist at? Within the Arcadia REPL?

**Comment by nasser** [comment]
@cjsauer yup, and make sure to replace `"/Applications/Unity/Unity.app/Contents/MonoBleedingEdge/lib/mono/4.5/mozroots.exe"` with the equivalent for you, it should be in a comparable place relative to your Unity installation on Windows. Let me know if it fixes it for you. Arcadia will do something like this automatically eventually.

**Comment by nasser** [comment]
Fixed as of 059d591
